src/Umbraco.Web.Common/ApplicationBuilder/UmbracoApplicationBuilder.cs

```diff
@@ -2,50 +2,130 @@
 using Microsoft.AspNetCore.Builder;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Options;
+using SixLabors.ImageSharp.Web.DependencyInjection;
 using Umbraco.Cms.Core.Services;
+using Umbraco.Extensions;
 
 namespace Umbraco.Cms.Web.Common.ApplicationBuilder
 {
     /// <summary>
     /// A builder to allow encapsulating the enabled endpoints in Umbraco
     /// </summary>
-    internal class UmbracoApplicationBuilder : IUmbracoApplicationBuilder, IUmbracoMiddlewareBuilder
+    public class UmbracoApplicationBuilder : IUmbracoApplicationBuilder, IUmbracoEndpointBuilder, IUmbracoApplicationBuilderContext
     {
-        public UmbracoApplicationBuilder(IServiceProvider services, IRuntimeState runtimeState, IApplicationBuilder appBuilder)
+        private Action<IUmbracoApplicationBuilderContext> _customMiddlewareRegistration;
+
+        public UmbracoApplicationBuilder(IApplicationBuilder appBuilder)
         {
-            ApplicationServices = services;
-            RuntimeState = runtimeState;
-            AppBuilder = appBuilder;
+            ApplicationServices = appBuilder.ApplicationServices;
+            RuntimeState = appBuilder.ApplicationServices.GetRequiredService<IRuntimeState>();
+            AppBuilder = appBuilder ?? throw new ArgumentNullException(nameof(appBuilder));
```

Comment:
This null check doesn't look quite right... as if `appBuilder` were null we'd get exceptions on the two lines above.  Just need to move the `AppBuilder` property setting first in the method.